matchzoo/contrib/layers/matching_tensor_layer.py

```diff
@@ -0,0 +1,137 @@
+import typing
+
+import numpy as np
+
+from keras import backend as K
+from keras.engine import Layer
+from keras.initializers import constant
+
+
+class MatchingTensorLayer(Layer):
+    """
+    Layer that captures the basic interactions between two tensors.
+
+    :param channels: Number of word interaction tensor channels
+    :param normalize: Whether to L2-normalize samples along the
+        dot product axis before taking the dot product.
+        If set to True, then the output of the dot product
+        is the cosine proximity between the two samples.
+    :param init_diag: Whether to initialize the diagonal elements
+        of the matrix.
+    :param kwargs: Standard layer keyword arguments.
+
+    Examples:
+        >>> import matchzoo as mz
+        >>> layer = mz.contrib.layers.MatchingTensorLayer(channels=4,
+        ...                                               normalize=True,
+        ...                                               init_diag=True)
+        >>> num_batch, left_len, right_len, num_dim = 5, 3, 2, 10
+        >>> layer.build([[num_batch, left_len, num_dim],
+        ...              [num_batch, right_len, num_dim]])
+
+    """
+
+    def __init__(self, channels: int = 4, normalize: bool = True,
+                 init_diag: bool = True, **kwargs):
+        """:class:`MatchingTensorLayer` constructor."""
+        super(MatchingTensorLayer, self).__init__(**kwargs)
+        self._channels = channels
+        self._normalize = normalize
+        self._init_diag = init_diag
+        self._shape1 = None
+        self._shape2 = None
+
+    def build(self, input_shape: list):
+        """
+        Build the layer.
+
+        :param input_shape: the shapes of the input tensors,
+            for MatchingTensorLayer we need two input tensors.
+        """
+        # Used purely for shape validation.
+        if not isinstance(input_shape, list) or len(input_shape) != 2:
+            raise ValueError('A `MatchingTensorLayer` layer should be called '
+                             'on a list of 2 inputs.')
+        self._shape1 = input_shape[0]
+        self._shape2 = input_shape[1]
+        for idx in (0, 2):
+            if self._shape1[idx] != self._shape2[idx]:
+                raise ValueError(
+                    'Incompatible dimensions: '
+                    f'{self._shape1[idx]} != {self._shape2[idx]}.'
+                    f'Layer shapes: {self._shape1}, {self._shape2}.'
+                )
+
+        if self._init_diag:
+            if self._shape1[2] != self._shape2[2]:
+                raise ValueError('Use init_diag need same embedding shape.')
+            M_diag = np.float32(
+                np.random.uniform(
+                    -0.05, 0.05,
+                    [self._channels, self._shape1[2], self._shape2[2]]
+                )
+            )
+            for i in range(self._channels):
+                for j in range(self._shape1[2]):
+                    M_diag[i][j][j] = 0.1
```

Comment:
if I were you, I would do:

```suggestion
for channel_index in range(self._channels):
    np.fill_diagonal(m_diag[channel_index], 0.1)
```

Reply:
Thanks for your suggestion.